Honor properties sortOrder in the edit dialog

josemarcenaro opened the pull request:
The reason for this addition is that when a document type has inherited properties, the sort order in the form ... is a mess.

A method is added to the controller and resource to fetch the sortOrder
from the Content Type and assign it to each scaffolded property.

Then the sortOrder is used for sorting the form fields:
 ng-repeat="property in tab.properties | orderBy:'+sortOrder'"

Credit to Sentient for the idea of pulling this trick, in a slightly different context:
https://our.umbraco.org/forum/umbraco-7/using-umbraco-7/70292-sort-order-problems-with-inherited-document-type-properties

mattbrailsford commented on src/Our.Umbraco.DocTypeGridEditor/Web/UI/App_Plugins/DocTypeGridEditor/Js/doctypegrideditor.resources.js:
```diff
@@ -51,6 +51,9 @@
                     }
                 });
+            },
+            getContentTypePropertyTypes: function (alias) {
```
Should be wrapped in umbRequestHelper for consistency.

josemarcenaro replied:
Oops, you are right.  Done in new commit.

mattbrailsford commented on src/Our.Umbraco.DocTypeGridEditor/Web/Controllers/DocTypeGridEditorApiController.cs:
```diff
@@ -83,5 +84,35 @@ public object GetDataTypePreValues(string dtdId)
+            if (doctype != null)
+            {
+                list = doctype.PropertyTypes.ToList();
+                if (doctype.ParentId != 0)
```
How does this work with compositions where they have multiple "parents"? Does PripertyTypes contain composition property types? Need to investigate.

josemarcenaro replied:
You are right, something smells bad here.
Looks like if we need the sort order of the inherited properties we should be checking the CompositionPropertyGroups instead of the ParentId.

I will investigate and get back to you.

mattbrailsford commented on src/Our.Umbraco.DocTypeGridEditor/Web/UI/App_Plugins/DocTypeGridEditor/Js/doctypegrideditor.controllers.js:
```diff
@@ -211,6 +211,19 @@ angular.module("umbraco").controller("Our.Umbraco.DocTypeGridEditor.Dialogs.DocT
+                    }, function (error) { });
+
                     // Assign the model to scope
                     $scope.nodeContext = $scope.node = data;
```
Not sure if this need to be inside the getContentTypePropertyTypes then statement?

mattbrailsford commented on src/Our.Umbraco.DocTypeGridEditor/Web/UI/App_Plugins/DocTypeGridEditor/Views/doctypegrideditor.dialog.html:
```diff
@@ -15,7 +15,7 @@
 
                     <umb-property property="property"
-                                  ng-repeat="property in tab.properties">
+                                  ng-repeat="property in tab.properties | orderBy:'+sortOrder'">
```
I seem to recall that the OrderBy filter isn't that performant. I'd be tempted to sort them using in the controller before they are bound.

josemarcenaro replied:
I didn't think performance on the sorting of a - usually short - list of fields wouldn't be an issue. But I can change to sort in the controller if you consider it might be.

josemarcenaro commented:
Matt, thanks for your prompt review of this PR.
I need to further investigate the composition behavior.  Hold on.

josemarcenaro commented:
Matt, after reviewing your comments I rewrote the method I had added to the DocTypeGridEditorApiController (server side) to take into account the **compositions** and the tabs (CompositionGroups) in each of them.

Once I did that, I came to a disturbing conclusion: ***that's exactly the order in which contentResource.getScaffold() returns the property list***.

So it looks like this PR makes no sense - Please disregard it.

Question: should I close it or you´ll do it? 
Thanks for your review and comments.
Jose

mattbrailsford commented:
No worries, thanks for doing a thorough check, really appreciated.
